### ISSUE TITLE
How do I turn off these warnings?

### ISSUE THREAD

**Issue by lintao185** [dotnet/TorchSharp#1322]
![image](https://github.com/dotnet/TorchSharp/assets/55724885/dd232aca-16bb-4742-9a5b-fae5f5a704dc)
Excuse me, how do I turn off these warnings?

**Comment by yueyinqiu** [comment]
may related to #1307

Hmm... I've checked it again. PyTorch is checking those things in the Python codes, and warning with `warnings` package. But why LibTorch is doing that at C++ level and warning by printing to `stderr`? That's pretty strange...

By the way, in most cases, you should resolve those warnings, instead of just ignoring or suppressing them.

**Comment by yueyinqiu** [comment]
Actually we can set the warning handler of `c10` to redirect or 'catch' the warnings. But what should happen after the warnings have been catched? C# does not have a standard behavior for it... @NiklasGustafsson

**Comment by lintao185** [comment]
> may related to #1307
> 
> Hmm... I've checked it again. PyTorch is checking those things in the Python codes, and warning with `warnings` package. But why LibTorch is also doing that at C++ level and warning with `c10`? That's pretty strange...
> 
> By the way, in most cases, you should resolve those warnings, instead of just ignoring or suppressing them.

![image](https://github.com/dotnet/TorchSharp/assets/55724885/b2e23559-e25a-45e6-b80a-5b50175df3a5)
I just pushed the model to the GPU, which is a bit strange.

**Comment by yueyinqiu** [comment]
Hmm... It seems to happen because `_toEpilog` is copying `param.grad` without checking that in advance.

**Comment by lintao185** [comment]
![image](https://github.com/dotnet/TorchSharp/assets/55724885/7fa40b28-e068-422e-9023-e57c7ca6e68b)
`.zero_grad()` will also cause warnings.
![image](https://github.com/dotnet/TorchSharp/assets/55724885/c6e3c7c3-25bd-427f-b337-c5cc22ac1a17)
` this.conv.weight[..] = nn.Parameter(x.view(1, c1, 1, 1));`
The root cause has been found, and changing the above code to the following resolves the issue.
```
 using (torch.no_grad())
 {
     this.conv.weight[..] = nn.Parameter(x.view(1, c1, 1, 1));
 }
```

**Comment by yueyinqiu** [comment]
Hmmm... That's strange. In my expectation `torch.no_grad` should do nothing with that.

Anyway could you provide the implemention if `RequiresGrad` so I can reproduce it?

**Comment by lintao185** [comment]
public static class ModuleExtend
{
    /// <summary>
    /// 设置模型参数是否需要更新
    /// </summary>
    /// <param name="module">模型</param>
    /// <param name="requires_grad">是否需要斜率</param>
    /// <returns>模型</returns>
    public static T RequiresGrad<T>(this T module,bool requires_grad) where T: nn.Module
    {
        foreach (var parameter in module.parameters())
        {
            parameter.requires_grad_(requires_grad);
        }

        return module;
    }
}

**Comment by yueyinqiu** [comment]
Ahhh? That seems to mean:

```csharp
using TorchSharp;

var tensor = torch.zeros([1]);
var tensor2 = torch.zeros([1], requires_grad: true);
Console.WriteLine(tensor.requires_grad); // False
tensor[..] = tensor2;
Console.WriteLine(tensor.requires_grad); // True
```

...Hmmm....

**Comment by yueyinqiu** [comment]
Well PyTorch has the same behavior... I don't know...

```python
import torch

tensor = torch.zeros([1])
tensor2 = torch.zeros([1], requires_grad=True)
tensor[:] = tensor2
print(tensor.requires_grad)  # True
```

**Comment by yueyinqiu** [comment]
A minial reproduction might be this:

```csharp
using TorchSharp;
using static TorchSharp.torch;

var p = nn.Parameter(torch.zeros([1]), requires_grad: false);
p[..] = nn.Parameter(torch.zeros([1]));

Console.WriteLine(p.requires_grad);
Console.WriteLine(p.grad());
```

It prints

```plain
True
[W TensorBody.h:494] Warning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). If you indeed want the .grad field to be populated for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See github.com/pytorch/pytorch/pull/30531 for more informations. (function grad)
```

**Comment by lintao185** [comment]
Yes

**Comment by yueyinqiu** [comment]
It seems to be a problem of LibTorch. PyTorch has the same problem:

```python
import torch

p = torch.nn.Parameter(torch.zeros([1]), requires_grad=False)
p[:] = torch.zeros([1], requires_grad=True)
print(p.requires_grad)
print(p.grad)
```

It prints (due to some reason of PyCharm, the order is not kept):

```python
UserWarning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). If you indeed want the .grad field to be populated for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See github.com/pytorch/pytorch/pull/30531 for more informations. (Triggered internally at aten\src\ATen/core/TensorBody.h:494.)
  print(p.grad)
True
None
```

**Comment by lintao185** [comment]
```
    p = nn.Parameter(torch.ones(1))
    p2 = nn.Parameter(torch.ones(1))
    p.requires_grad_(False)
    print(p.requires_grad) #False
    print(p2.requires_grad)#True
    p.data[:]=p2
    print(p.requires_grad)#False
```
```
var p = nn.Parameter(torch.ones(1));
var p2 = nn.Parameter(torch.ones(1));
p.requires_grad_(false);
Console.WriteLine(p.requires_grad);//False
Console.WriteLine(p2.requires_grad);//True
p[..] = p2;
Console.WriteLine(p.requires_grad);//True
```

**Comment by yueyinqiu** [comment]
Oh yes so that's not a bug but a intentional design. And the `xxx.weight[:] = xxx` is making the weight to be a non-leaf tensor.

**Comment by lintao185** [comment]
The behaviors of PyTorch and TorchSharp are somewhat inconsistent.

**Comment by yueyinqiu** [comment]
remove '.data' in pytorch and then they are the same.

if you want to do it in torchsharp, try detach.

**Comment by NiklasGustafsson** [comment]
@lintao185 -- given what @yueyinqiu said in https://github.com/dotnet/TorchSharp/issues/1323#issuecomment-2166999264, should this be closed, too?
